gcloud/search/document.py

```diff
@@ -0,0 +1,303 @@
+        return client
+
+    def _build_value_resource(self, value):
+        """Helper for _build_fields_resource"""
+        result = {}
+        if value.value_type == 'string':
+            result['stringValue'] = value.string_value
+            if value.string_format is not None:
+                result['stringFormat'] = value.string_format
+            if value.language is not None:
+                result['lang'] = value.language
+        elif value.value_type == 'number':
+            result['numberValue'] = value.number_value
+        elif value.value_type == 'timestamp':
+            stamp = value.timestamp_value.strftime(_RFC3339_MICROS)
+            result['timestampValue'] = stamp
+        elif value.value_type == 'geo':
+            result['geoValue'] = '%s, %s' % value.geo_value
+        else:
+            raise ValueError('Unknown value_type: %s' % value.value_type)
+        return result
+
+    def _build_fields_resource(self):
+        """Helper for create"""
+        fields = {}
+        for field_name, field in self.fields.items():
+            if field.values:
+                values = []
+                fields[field_name] = {'values': values}
+                for value in field.values:
+                    values.append(self._build_value_resource(value))
+        return fields
+
+    def _set_properties(self, api_response):
+        """Helper for create, reload"""
+        self.rank = api_response.get('rank')
+        self._parse_fields_resource(api_response)
+
+    def create(self, client=None):
+        """API call:  create the document via a PUT request
+
+        See:
+        https://cloud.google.com/search/reference/rest/v1/projects/indexes/documents/create
+
+        :type client: :class:`gcloud.search.client.Client` or ``NoneType``
+        :param client: the client to use.  If not passed, falls back to the
+                       ``client`` stored on the current document's index.
+        """
+        data = {'docId': self.name}
+
+        if self.rank is not None:
+            data['rank'] = self.rank
+
+        fields = self._build_fields_resource()
+        if fields:
+            data['fields'] = fields
+
+        client = self._require_client(client)
+        api_response = client.connection.api_request(
+            method='PUT', path=self.path, data=data)
+
+        self._set_properties(api_response)
+
+    def exists(self, client=None):
+        """API call:  test existence of the document via a GET request
+
+        See
+        https://cloud.google.com/search/reference/rest/v1/projects/indexes/documents/get
+
+        :type client: :class:`gcloud.search.client.Client` or ``NoneType``
+        :param client: the client to use.  If not passed, falls back to the
+                       ``client`` stored on the current document's index.
+        """
+        client = self._require_client(client)
+        try:
+            client.connection.api_request(method='GET', path=self.path)
```

Comment:
Maybe the 404 is because the API is still "private"?  The apis-explorer site doesn't show any entry for "cloud search".